File: 1726.tuple-with-same-product.cpp
#include <numeric>
#include <vector>
#include <unordered_map>

using namespace std;

// @leet start
class Solution {
public:
    int tupleSameProduct(vector<int>& nums) {
        std::unordered_map<int, int> products;
        products.reserve(nums.size() * (nums.size() - 1) / 2);
        for (int i = 1; i < nums.size(); ++i) {
            for (int j = 0; j < i; ++j) {
                ++products[nums[i] * nums[j]];
            }
        }
        return std::accumulate(products.begin(), products.end(), 0, [](int a, const auto &p) {
            return a + p.second * (p.second - 1) * 4;
        });
    }
};
// @leet end

File: 1726.tuple-with-same-product.rs
// @leet start
use std::collections::HashMap;

impl Solution {
    pub fn tuple_same_product(nums: Vec<i32>) -> i32 {
        let mut products = HashMap::new();
        for i in 1 .. nums.len() {
            for j in 0 .. i {
                products.entry(nums[i] * nums[j]).and_modify(|c| *c += 1).or_insert(1);
            }
        }
        let mut count = 0;
        products.iter().for_each(|(&p, &c)| {
            count += 4 * c * (c - 1);
        });
        count
    }
}
// @leet end
